### PR TITLE
fix(retrofit): Explicitly mark BAD_REQUEST and NOT_FOUND errors as not retryable

### PR DESCRIPTION
**Pull request by dbyron0**
cc: @ajordens .. an attempt to address https://github.com/spinnaker/spinnaker/issues/6237

### CODE REVIEW

File: kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/exceptions/SpinnakerRetrofitErrorHandler.java [hunk 1]
```diff
@@ -52,9 +52,14 @@ public Throwable handleError(RetrofitError e) {
     switch (e.getKind()) {
       case HTTP:
         if (e.getResponse().getStatus() == HttpStatus.NOT_FOUND.value()) {
-          return new NotFoundException(e);
+          return new NotFoundException(e).setRetryable(false);
```
Comment (dbyron0) [inline kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/exceptions/SpinnakerRetrofitErrorHandler.java:55]: Better to change NotFoundException to always set retryable to false?

**Review by link108 — APPROVED** [review]
LGTM

File: kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/exceptions/SpinnakerRetrofitErrorHandler.java [hunk 2]
```diff
@@ -52,9 +52,14 @@ public Throwable handleError(RetrofitError e) {
...
         }
-        return new SpinnakerHttpException(e);
+        SpinnakerHttpException retval = new SpinnakerHttpException(e);
+        if (HttpStatus.Series.resolve(e.getResponse().getStatus())
+            == HttpStatus.Series.CLIENT_ERROR) {
+          retval.setRetryable(false);
```
Comment (ajordens) [inline kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/exceptions/SpinnakerRetrofitErrorHandler.java:60]: How much testing has been done with this outside of `kork`?  At a glance `Series.CLIENT_ERROR` is any 4xx, including 429 which could be retryable.

This does seem like a long drawn out fix for something that could have just been done in rosco? Sorry the holiday period made the cycle time on this quite long.

```
Response response =
        retrySupport.retry(
            () ->
                AuthenticatedRequest.allowAnonymous(
                    () -> clouddriverService.fetchArtifact(artifact)),
            5,
            1000,
            true);
```

Why can't we just wrap `clouddriverService.fetchArtifact(artifact))` in a try/catch that checks the status code?

Comment (dbyron0) [reply kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/exceptions/SpinnakerRetrofitErrorHandler.java:60]: I haven't tested this outside of kork, and you're right about not handling 429 properly.

You're also right that adding the try/catch block to rosco is the most direct/simplest way forward.  It does seem like the kind of thing to handle centrally though.  I haven't come across other places that would benefit from it yet, but I bet they exist.

What do you think of keeping the change here, but limiting it to 400/bad request?

Comment (dbyron0) [reply kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/exceptions/SpinnakerRetrofitErrorHandler.java:59]: Like this: https://github.com/spinnaker/kork/pull/837/commits/11a5ff9363824d427ff99efc9ab68f0d76815f80

Comment (ajordens) [reply kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/exceptions/SpinnakerRetrofitErrorHandler.java:59]: That looks more reasonable to me. I'll set a reminder to take one last look before merging tomorrow.

**Pull request merged**